lib/core/src/Cardano/Wallet.hs

```diff
@@ -1759,27 +1794,61 @@ balanceTransactionWithSelectionStrategy
         let minfee = evaluateMinimumFee tl nodePParams tx
         let update = TxUpdate [] [] [] (UseNewTxFee minfee)
         tx' <- left ErrBalanceTxUpdateError $ updateTx tl tx update
-        let balance = evaluateTransactionBalance tl tx' nodePParams
-                (CS.toExternalUTxOMap $ UTxOIndex.toMap internalUtxoAvailable)
-                (view #inputs ptx)
+        let balance = evaluateTransactionBalance tl tx' nodePParams combinedUTxO
         let minfee' = Cardano.Lovelace $ fromIntegral $ unCoin minfee
         return (balance, minfee')
 
+    -- | Ensure the wallet UTxO is consistent with a provided @Cardano.UTxO@.
+    --
+    -- They are not consistent iff an input can be looked up in both UTxO sets
+    -- with different @Address@, or @TokenBundle@ values.
+    --
+    -- The @Cardano.UTxO era@ is allowed to contain additional information, like
+    -- datum hashes, which the wallet UTxO cannot represent.
+    --
+    -- NOTE: Representing the wallet utxo as a @Cardano.UTxO@ will not make this
+    -- check easier, even if it may be useful in other regards.
+    guardWalletUTxOConsistencyWith
+        :: Cardano.UTxO era
+        -> ExceptT ErrBalanceTx m ()
+    guardWalletUTxOConsistencyWith u' = do
+        let u = Map.mapKeys (fromCardanoTxIn tl)
+                . Map.map (fromCardanoTxOut tl)
+                $ (unUTxO u')
+        let conflicts = lefts $ flip map (Map.toList u) $ \(i, o) ->
+                case i `UTxO.lookup` walletUTxO of
+                    Just o' -> unless (o == o') $ Left (o, o')
+                    Nothing -> pure ()
+
+        unless (null conflicts) $
+            throwE ErrBalanceTxConflictingInputResolution
+      where
+         unUTxO (Cardano.UTxO u) = u
```

Comment:
Hi @Anviking!

I find this function a little difficult to understand, especially at first glance.

Though more importantly, I suspect it's difficult to test that the combining behaviour is correct in isolation, because we have to call it via `balanceTransaction`.

Given that we're guarding the correctness of `combinedUTxO`, perhaps we could extract out both the definition of `combinedUTxO` and `guardWalletUTxOConsistencyWith` into a separate **_pure_** function that is easy to test?

The function would take a pair of `Cardano.UTxO` values, and return a combined `Cardano.UTxO` if and only if there are no inconsistencies.

I've written a prototype here: 8866ffde6222944a69ba18a278816409e49920fa.

Since we don't care about reporting which values are in conflict (at the moment), we can use simple set operations to determine if there are conflicts.

Reply:
I hear you, and it did cross my mind, but I'm not sure.

My reservation is: it's incompatible with a global `combinedUTxO` `where` binding. It's crucial for helpers like `balanceAfterSettingMinFee`, `guardTxBalance`, `assembleTransaction` to use the same `Cardano.UTxO`. With a `combinedUTxO <- combineWalletUTxOAndExternalUTxO` we'd need to either manually pass the `utxo` around, which adds noise and room for mistakes, or add more indirection like increased nesting or an internal "Balance monad". The latter I'm curious about, but now's the wrong time to think about it.

Extended prototype showing how it would turn out -> 2eaab252d1a192b63a4ffdb8e92ccd4ecedc4d01.

> Though more importantly, I suspect it's difficult to test that the combining behaviour is correct in isolation, because we have to call it via balanceTransaction.

1. `prop_balanceTransactionValid` currently fails without `guardWalletUTxOConsistencyWith`. The guard was added because of a dropped [hack](https://github.com/input-output-hk/cardano-wallet/pull/3255/files#diff-45f59e6a8905a1e2f6cc28025e3c5e5ab6e37b9c79735f222569e677d5d99ca2L3362-L3375) which artificially made the test data consistent. (Sorry, this was not easily apparent from the commit history)
2. If possible, I would much prefer to focus future tests for this on the `balanceTransaction` interface. (E.g. if the script inputs I'm trying to redeem exist in my wallet utxo, I should still be able to redeem them.)

So thanks for raising this and for the prototype (I do want to move to Cardano.UTxO types like you did there), but could we ignore this for now?